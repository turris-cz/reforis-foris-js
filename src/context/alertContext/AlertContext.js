/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState, useContext, useCallback, useMemo } from "react";

import PropTypes from "prop-types";

import Alert, { ALERT_TYPES } from "../../bootstrap/Alert";
import Portal from "../../utils/Portal";

AlertContextProvider.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

function AlertContextProvider({ children }) {
    const { AlertContext } = window;
    const [alert, setAlert] = useState(null);

    const setAlertWrapper = useCallback(
        (message, type = ALERT_TYPES.DANGER) => {
            setAlert({ message, type });
        },
        [setAlert]
    );

    const dismissAlert = useCallback(() => setAlert(null), [setAlert]);
    const contextValue = useMemo(
        () => [setAlertWrapper, dismissAlert],
        [setAlertWrapper, dismissAlert]
    );

    return (
        <>
            {alert && (
                <Portal containerId="alert-container">
                    <Alert type={alert.type} onDismiss={dismissAlert}>
                        {alert.message}
                    </Alert>
                </Portal>
            )}
            <AlertContext.Provider value={contextValue}>
                {children}
            </AlertContext.Provider>
        </>
    );
}

function useAlert() {
    const { AlertContext } = window;
    return useContext(AlertContext);
}

export { AlertContextProvider, useAlert };
