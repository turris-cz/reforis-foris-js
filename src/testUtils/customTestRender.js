/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from "react";
import { UIDReset } from "react-uid";
import { StaticRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";

import { AlertContextMock } from "./alertContextMock";
import { CustomizationContextMock } from "./customizationContextMock";

Wrapper.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

function Wrapper({ children }) {
    return (
        <CustomizationContextMock>
            <AlertContextMock>
                <StaticRouter>
                    <UIDReset>{children}</UIDReset>
                </StaticRouter>
            </AlertContextMock>
        </CustomizationContextMock>
    );
}

const customTestRender = (ui, options) =>
    render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customTestRender as render };
