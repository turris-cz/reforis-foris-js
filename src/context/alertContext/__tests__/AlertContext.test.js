/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, queryByText, fireEvent } from "customTestRender";

import { useAlert, AlertContextProvider } from "../AlertContext";

function AlertTest() {
    const [setAlert, dismissAlert] = useAlert();
    // alert-container serves as an output for Portal which renders Alert
    return (
        <>
            <div id="alert-container" />
            <button onClick={() => setAlert("Alert content")}>Set alert</button>
            <button onClick={dismissAlert}>Dismiss alert</button>
        </>
    );
}

describe("AlertContext", () => {
    let componentContainer;

    beforeEach(() => {
        const { container } = render(
            <AlertContextProvider>
                <AlertTest />
            </AlertContextProvider>
        );
        componentContainer = container;
    });

    it("should render component without alert", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should render alert", () => {
        fireEvent.click(getByText(componentContainer, "Set alert"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should dismiss alert with alert button", async () => {
        fireEvent.click(getByText(componentContainer, "Set alert"));
        // Alert is present
        expect(getByText(componentContainer, "Alert content")).toBeDefined();

        fireEvent.click(componentContainer.querySelector(".btn-close"));
        // Alert is gone
        await (() =>
            expect(
                queryByText(componentContainer, "Alert content")
            ).toBeNull());
    });

    it("should dismiss alert with external button", () => {
        fireEvent.click(getByText(componentContainer, "Set alert"));
        // Alert is present
        expect(getByText(componentContainer, "Alert content")).toBeDefined();

        fireEvent.click(getByText(componentContainer, "Dismiss alert"));
        // Alert is gone
        expect(queryByText(componentContainer, "Alert content")).toBeNull();
    });
});
