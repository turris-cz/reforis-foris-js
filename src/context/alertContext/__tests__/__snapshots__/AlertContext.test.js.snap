// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertContext should render alert 1`] = `
<div>
  <div
    id="alert-container"
  >
    <div
      class="alert alert-danger alert-fade-in alert-dismissible"
      role="alert"
    >
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
      Alert content
    </div>
  </div>
  <button>
    Set alert
  </button>
  <button>
    Dismiss alert
  </button>
</div>
`;

exports[`AlertContext should render component without alert 1`] = `
<div>
  <div
    id="alert-container"
  />
  <button>
    Set alert
  </button>
  <button>
    Dismiss alert
  </button>
</div>
`;
